# Fuel Economy Data

In [1]:
import pandas as pd
import numpy as np

In [2]:
# load EPA fuel economy dataset downloaded from Kaggle
df = pd.read_csv('./data/database.csv')

C:\Users\chloe\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (7,16,62,64,65,67,68,78) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [8]:
makes = df['Make'].unique()

In [18]:
df.describe()

Vehicle ID          Year  Engine Index  Engine Cylinders  \
count  38113.000000  38113.000000  38113.000000      37977.000000   
mean   19170.638496   2000.194527   8799.389001          5.736656   
std    11134.878665     10.464573  17781.058490          1.752254   
min        1.000000   1984.000000      0.000000          2.000000   
25%     9529.000000   1991.000000      0.000000          4.000000   
50%    19058.000000   2001.000000    212.000000          6.000000   
75%    28779.000000   2009.000000   4451.000000          6.000000   
max    38542.000000   2017.000000  69102.000000         16.000000   

       Engine Displacement  City MPG (FT1)  Unrounded City MPG (FT1)  \
count         37979.000000    38113.000000              38113.000000   
mean              3.317583       17.981109                  4.606426   
std               1.361995        6.849728                 10.113963   
min               0.000000        6.000000                  0.000000   
25%               2.200000       15.000000                  0.000000   
50%               3.000000       17.000000                  0.000000   
75%               4.300000       20.000000                  0.000000   
max               8.400000      150.000000                150.000000   

       City MPG (FT2)  Unrounded City MPG (FT2)  \
count    38113.000000              38113.000000   
mean         0.546218                  0.386633   
std          4.109282                  3.861179   
min          0.000000                  0.000000   
25%          0.000000                  0.000000   
50%          0.000000                  0.000000   
75%          0.000000                  0.000000   
max        145.000000                145.083500   

       City Gasoline Consumption (CD)  ...  Hours to Charge (240V)  \
count                    38113.000000  ...            38113.000000   
mean                         0.000432  ...                0.027635   
std                          0.039976  ...                0.473685   
min                          0.000000  ...                0.000000   
25%                          0.000000  ...                0.000000   
50%                          0.000000  ...                0.000000   
75%                          0.000000  ...                0.000000   
max                          5.350000  ...               12.000000   

       Hours to Charge (AC 240V)  Composite City MPG  Composite Highway MPG  \
count               38113.000000        38113.000000           38113.000000   
mean                    0.005549            0.082203               0.080891   
std                     0.161014            2.156682               2.052187   
min                     0.000000            0.000000               0.000000   
25%                     0.000000            0.000000               0.000000   
50%                     0.000000            0.000000               0.000000   
75%                     0.000000            0.000000               0.000000   
max                     7.000000           97.000000              81.000000   

       Composite Combined MPG   Range (FT1)  City Range (FT1)  \
count            38113.000000  38113.000000      38113.000000   
mean                 0.081311      0.469708          0.426249   
std                  2.097794      9.352069          9.104702   
min                  0.000000      0.000000          0.000000   
25%                  0.000000      0.000000          0.000000   
50%                  0.000000      0.000000          0.000000   
75%                  0.000000      0.000000          0.000000   
max                 88.000000    315.000000        305.900000   

       Highway Range (FT1)  City Range (FT2)  Highway Range (FT2)  
count         38113.000000      38113.000000         38113.000000  
mean              0.419197          0.043973             0.040051  
std               9.315914          1.311628             1.169281  
min               0.000000          0.000000             0.000000  
25%      

In [11]:
makes

array(['Alfa Romeo', 'AM General', 'American Motors Corporation',
       'Aston Martin', 'Audi', 'Avanti Motor Corporation', 'Bertone',
       'Bill Dovell Motor Car Company', 'Bitter Gmbh and Co. Kg', 'BMW',
       'Buick', 'Cadillac', 'Chevrolet', 'Chrysler', 'Dodge', 'Ford',
       'GMC', 'Grumman Olson', 'Honda', 'Import Foreign Auto Sales Inc',
       'Isuzu', 'Jaguar', 'Jeep', 'Kenyon Corporation Of America',
       'Lamborghini', 'Lincoln', 'Lotus', 'Maserati', 'Mazda',
       'Mercedes-Benz', 'Mercury', 'Mitsubishi', 'Nissan', 'Oldsmobile',
       'Peugeot', 'Pininfarina', 'Plymouth', 'Pontiac', 'Porsche',
       'Renault', 'Rolls-Royce', 'S and S Coach Company  E.p. Dutton',
       'Saab', 'Subaru', 'Superior Coaches Div E.p. Dutton', 'Suzuki',
       'Toyota', 'Volkswagen', 'Volvo', 'E. P. Dutton, Inc.', 'Ferrari',
       'Merkur', 'TVR Engineering Ltd', 'Acura', 'Autokraft Limited',
       'Hyundai', 'Lambda Control Systems', 'London Coach Co Inc',
       'Panther Car Compan

In [17]:
df[df.isnull().any(axis=1)]

Vehicle ID  Year        Make               Model  \
0           26587  1984  Alfa Romeo           GT V6 2.5   
1           27705  1984  Alfa Romeo           GT V6 2.5   
2           26561  1984  Alfa Romeo  Spider Veloce 2000   
3           27681  1984  Alfa Romeo  Spider Veloce 2000   
4           27550  1984  AM General   DJ Po Vehicle 2WD   
...           ...   ...         ...                 ...   
38108       37564  2017       Volvo            XC60 FWD   
38109       37547  2017       Volvo            XC90 AWD   
38110       37548  2017       Volvo            XC90 AWD   
38111       37703  2017       Volvo       XC90 AWD PHEV   
38112       37687  2017       Volvo            XC90 FWD   

                                    Class              Drive  \
0                        Minicompact Cars                NaN   
1                        Minicompact Cars                NaN   
2                             Two Seaters                NaN   
3                             Two Seaters                NaN   
4             Special Purpose Vehicle 2WD      2-Wheel Drive   
...                                   ...                ...   
38108     Small Sport Utility Vehicle 2WD  Front-Wheel Drive   
38109  Standard Sport Utility Vehicle 4WD    All-Wheel Drive   
38110  Standard Sport Utility Vehicle 4WD    All-Wheel Drive   
38111  Standard Sport Utility Vehicle 4WD    All-Wheel Drive   
38112  Standard Sport Utility Vehicle 2WD  Front-Wheel Drive   

            Transmission Transmission Descriptor  Engine Index  \
0         Manual 5-Speed                     NaN          9001   
1         Manual 5-Speed                     NaN          9005   
2         Manual 5-Speed                     NaN          9002   
3         Manual 5-Speed                     NaN          9006   
4      Automatic 3-Speed                     NaN          1830   
...                  ...                     ...           ...   
38108     Automatic (S8)                     NaN            90   
38109     Automatic (S8)                     NaN            52   
38110     Automatic (S8)                     NaN            53   
38111     Automatic (S8)                     NaN            54   
38112     Automatic (S8)                     NaN            50   

      Engine Descriptor  ...  Hours to Charge (AC 240V)  Composite City MPG  \
0                 (FFS)  ...                        0.0                   0   
1        (FFS) CA model  ...                        0.0                   0   
2                 (FFS)  ...                        0.0                   0   
3        (FFS) CA model  ...                        0.0                   0   
4                 (FFS)  ...                        0.0                   0   
...                 ...  ...                        ...                 ...   
38108              SIDI  ...                        0.0                   0   
38109              SIDI  ...                        0.0                   0   
38110              SIDI  ...                        0.0                   0   
38111        SIDI; PHEV  ...                        0.0                  29   
38112              SIDI  ...                        0.0                   0   

      Composite Highway MPG Composite Combined MPG Range (FT1)  \
0                         0                      0           0   
1                         0                      0           0   
2                         0                      0           0   
3                         0                      0           0   
4                         0                      0           0   
...                     ...                    ...         ...   
38108                     0                      0           0   
38109                     0                      0           0   
38110                     0                      0           0   
38111                    32                     30           0   
38112                     0                      0           

There are many null values in this dataset. I will examine the dataset column by column to see what reasonable values for the different nulls should be.

In [58]:
def print_nulls(df):
    for column in df.columns:
        print(f"{column} Nulls: {df[df[column].isnull()].shape[0]}")
        print(f"{column} Valids: {df[df[column].notnull()].shape[0]}")

def nulls_to_df(df):
    """Returns a dataframe with a count of null and valid entries for each column in the
    original dataset."""
    d = dict()
    for column in df.columns:
        l = []
        l.append(df[df[column].isnull()].shape[0])
        l.append(df[df[column].notnull()].shape[0])
        d[column] = l
    
    new_df = pd.DataFrame.from_dict(d, orient='index', columns=['Nulls', 'Valids'])
    return new_df

In [35]:
transmissions = df['Transmission Descriptor'].unique()
transmissions

array([nan, 'Creeper', 'SIL', 'Overdrive', 'Lockup', 'VLKUP', '2MODE',
       '2MODE 2LKUP', '3LKUP', '2LKUP', 'CLKUP', '2MODE 3LKUP',
       'fuel injection', 'SIL Creeper', 'DC/FW', 'Lockup A3',
       '2MODE CLKUP', 'EMS 2MODE CLKU', 'EMS', '3MODE VLKUP', '4MODE',
       '3MODE CLKUP', 'SIL CMODE', '2MODE VLKUP', 'EMS 2MODE CLKUP',
       'EMS 5MODE', 'EMS 2MODE', '2MODE DC/FW', 'Elec Overdrive',
       'Mech Overdrive', '3MODE', 'CMODE', 'EMS 3MODE', '2MODE CLKUP FW',
       '3MODE 2LKUP', 'VMODE CLKUP', 'EMS CLKUP', 'EMS CMODE CLKUP',
       'SIL CLKUP', 'CMODE CLKUP', 'VMODE', 'EMS 3MODE CLKUP',
       '4MODE CLKUP', '6MODE', 'SIL EMS', 'VMODE VLKUP', '6MODE CLKUP',
       'SIL 3MODE', 'LONG RATIO', 'CMODE VLKUP', 'SIL 3MODE CLKUP',
       '3MODE CLKUP FW', 'SIL 2MODE CLKUP'], dtype=object)

In [49]:
print_nulls(df)

Vehicle ID Nulls: 0
Vehicle ID Valids: 38113
Year Nulls: 0
Year Valids: 38113
Make Nulls: 0
Make Valids: 38113
Model Nulls: 0
Model Valids: 38113
Class Nulls: 0
Class Valids: 38113
Drive Nulls: 1189
Drive Valids: 36924
Transmission Nulls: 11
Transmission Valids: 38102
Transmission Descriptor Nulls: 23068
Transmission Descriptor Valids: 15045
Engine Index Nulls: 0
Engine Index Valids: 38113
Engine Descriptor Nulls: 15420
Engine Descriptor Valids: 22693
Engine Cylinders Nulls: 136
Engine Cylinders Valids: 37977
Engine Displacement Nulls: 134
Engine Displacement Valids: 37979
Turbocharger Nulls: 32874
Turbocharger Valids: 5239
Supercharger Nulls: 37420
Supercharger Valids: 693
Fuel Type Nulls: 0
Fuel Type Valids: 38113
Fuel Type 1 Nulls: 0
Fuel Type 1 Valids: 38113
Fuel Type 2 Nulls: 36675
Fuel Type 2 Valids: 1438
City MPG (FT1) Nulls: 0
City MPG (FT1) Valids: 38113
Unrounded City MPG (FT1) Nulls: 0
Unrounded City MPG (FT1) Valids: 38113
City MPG (FT2) Nulls: 0
City MPG (FT2) Valids: 3811

In [64]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Alternate Charger                38067      46
Vehicle Charger                  38065      48
Electric Motor                   37509     604
Supercharger                     37420     693
Range (FT2)                      36680    1433
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

With 38,067 null values and only 46 valid values, **Alternate Charger** is in the worst shape, so I'll start with that. Based on the values in the column and on information found on the [Department of Energy's website](https://www.fueleconomy.gov/feg/ws/), this column seems to describe alternate chargers for electric vehicles. This does not apply to the vast majority of vehicles, so a reasonable value to substitute is "None".

In [65]:
df['Alternate Charger'].unique()

array([nan, '6.6 kW charger', 'dual charger', '80 amp dual charger',
       '3.6 kW charger'], dtype=object)

In [67]:
df['Alternate Charger'].fillna('None', inplace=True)
df['Alternate Charger'].unique()

array(['None', '6.6 kW charger', 'dual charger', '80 amp dual charger',
       '3.6 kW charger'], dtype=object)

In [69]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Vehicle Charger                  38065      48
Electric Motor                   37509     604
Supercharger                     37420     693
Range (FT2)                      36680    1433
Fuel Type 2                      36675    1438
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

I will clean the remaining columns similarly, relying on information from the US Department of Energy as well as column values to determine appropriate substitutions for nulls.

In [70]:
df['Vehicle Charger'].unique()

array([nan, '3.6 kW charger', 'single charger', 'standard charger',
       '6.6 kW charger', '7.2 kW charger'], dtype=object)

Vehicle charger also refers only to electric vehicles. Using the same logics as with Alternate Charger, I will subtitute these values with "None" also:

In [71]:
df['Vehicle Charger'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Electric Motor                   37509     604
Supercharger                     37420     693
Range (FT2)                      36680    1433
Fuel Type 2                      36675    1438
Gas Guzzler Tax                  35817    2296
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

In [72]:
df['Electric Motor'].unique()

array([nan, '85 kW AC Induction', '49 kW DC Brushless',
       '56kW AC Induction', '67 KW AC  Induction', '67 KW AC',
       '102kW AC Induction', '49kW DC Brushless', '62 KW AC Induction',
       '50 KW DC', '67 KW AC Induction', '27 KW AC Induction',
       '24 KW AC Synchronous', '36V Ni-MH', '288V Ni-MH', '330V Ni-MH',
       '158V Ni-MH', '150 kW', '245V Ni-MH', '202V Ni-MH', '300V Ni-MH',
       '312V Ni-MH', '275V Ni-MH', '101V Ni-MH', '126V Li-Ion',
       '125 kW AC Induction', '111 kW', '270V Li-Ion', '80 kW DCPM',
       '30 kW DCPM', '52 kW AC Induction', '374V Li-Ion', '115V Li-Ion',
       '75 kW AC PMSM', '100 kW DCPM', '2 @ 150 kw (300 kw)',
       '107 kW AC Induction', '144V Li-Ion', '346V Li-Ion', '49 kW DCPM',
       '260 kW AC Induction', '144V Ni-MH', '18 kW',
       '115 kW AC Induction', '266V Li-Ion', '82 kW ACIPM', '68 kW',
       '280V Li-Ion', '107 kW AC PMSM', '92 kW DCPM', '55 kW DCPM',
       '225 kW AC Induction', '270 kW AC Induction', '220V Li-Ion',
 

The **Electric Motor** column describes the kilowatt hours for the vehicle's electric motor. This doesn't apply to most vehicles, so substituting "None" is appropriate.

In [73]:
df['Electric Motor'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Supercharger                     37420     693
Range (FT2)                      36680    1433
Fuel Type 2                      36675    1438
Gas Guzzler Tax                  35817    2296
Alternative Fuel/Technology      35066    3047
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

In [74]:
df['Supercharger'].unique()

array([nan, 'S'], dtype=object)

The **Supercharger** column is different from what we've seen so far. It contains an "S" if the vehicle is supercharged and nothing otherwise. I think substituting "NS" (denoting "not supercharged") is a reasonable thing to do here:

In [75]:
df['Supercharger'].fillna('NS', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Range (FT2)                        36680    1433
Fuel Type 2                        36675    1438
Gas Guzzler Tax                    35817    2296
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [77]:
df['Range (FT2)'].unique()

array([nan, '360', '300', '230/270/270', '240/290/290', '220/270/260',
       '240/290/280', '290', '230/270', '240/280', '220/260', '140',
       '310', '280', '230', '210', '260', '120', '250/220/320', '270',
       '110', '290/340', '320', '310/390', '280/380', '130', '180',
       '270/240/340', '170', '310-540', '310/460', '310/540', '310/330',
       '310/480', '310/440', '330', '340-400', '370', '340', '340-440',
       '310-410', '240', '340-350', '290/410', '310/340', '250/370',
       '230/350', '240-420', '250', '310/320', '280/370', '350-370',
       '280-350', '310-420', '310-340', '330-450', '300-420', '310/410',
       '380', '410', '350', '310/420', '330/450', '220', '260/320',
       '310/370', '214', '236', '279', '312', '333', '307', '304', '383',
       '353', '413', '295', '294', '344', '338', '277', '308', '264',
       '303', '307/420', '337/451', '347', '286/382', '329', '35', '262',
       '292', '219', '354', '265', '366', '331', '245', '282', '337',
       '6

In [96]:
df[df['Range (FT2)'].notnull()][['Range (FT2)', 
                                 'City Range (FT2)', 'Highway Range (FT2)']].sort_values(
    by='Range (FT2)', ascending=False)

Range (FT2)  City Range (FT2)  Highway Range (FT2)
37015          97           103.030               90.550
34522          72            77.504               65.632
35769          72            77.504               65.632
33304          72            77.504               65.632
30089          69             0.000                0.000
...           ...               ...                  ...
33078          11            12.150                9.100
31911          11            12.150                9.100
35539          11            12.150                9.100
36491          11            11.370               11.190
34284          11            12.150                9.100

[1433 rows x 3 columns]

In [97]:
df.dtypes

Vehicle ID               int64
Year                     int64
Make                    object
Model                   object
Class                   object
                        ...   
City Range (FT1)       float64
Highway Range (FT1)    float64
Range (FT2)             object
City Range (FT2)       float64
Highway Range (FT2)    float64
Length: 81, dtype: object

**Range (FT2)** is trickier. The column denotes the range of vehicles using their second fuel type (FT2 = fuel type 2). The integer-like values of this column are an average of the City Range (FT2) and Highway Range (FT2) columns. However, inspecting the datatypes of this dataframe, I can see that there are also some string values in this column. I need to find these, make sense of them, and, if necessary, convert them to integers. 

In [162]:
df[df['Range (FT2)'].apply(type) == str]

Vehicle ID  Year       Make                                    Model  \
18087       16403  2000  Chevrolet                   Cavalier (Bi-fuel CNG)   
18162       16422  2000   Chrysler  Town and Country/Voyager/Grand Voy. 2WD   
18189       16421  2000      Dodge                Caravan/Grand Caravan 2WD   
18291       16411  2000       Ford                           Ranger 2WD FFV   
18292       16412  2000       Ford                           Ranger 2WD FFV   
...           ...   ...        ...                                      ...   
38033       38531  2017     Toyota                              Prius Prime   
38042       38291  2017     Toyota                          Sequoia 4WD FFV   
38054       38122  2017     Toyota                           Tundra 2WD FFV   
38057       38127  2017     Toyota                           Tundra 4WD FFV   
38111       37703  2017      Volvo                            XC90 AWD PHEV   

                                    Class                    Drive  \
18087                     Subcompact Cars        Front-Wheel Drive   
18162                       Minivan - 2WD        Front-Wheel Drive   
18189                       Minivan - 2WD        Front-Wheel Drive   
18291          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
18292          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
...                                   ...                      ...   
38033                        Midsize Cars        Front-Wheel Drive   
38042  Standard Sport Utility Vehicle 4WD  Part-time 4-Wheel Drive   
38054          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
38057          Standard Pickup Trucks 4WD  Part-time 4-Wheel Drive   
38111  Standard Sport Utility Vehicle 4WD          All-Wheel Drive   

                           Transmission Transmission Descriptor  Engine Index  \
18087                 Automatic 3-Speed                     NaN             0   
18162                 Automatic 4-Speed                   CLKUP             0   
18189                 Automatic 4-Speed                   CLKUP             0   
18291                 Automatic 4-Speed                   CLKUP             0   
18292                    Manual 5-Speed                     NaN             0   
...                                 ...                     ...           ...   
38033  Automatic (variable gear ratios)                     NaN           108   
38042                    Automatic (S6)                     NaN             8   
38054                    Automatic (S6)                     NaN             9   
38057                    Automatic (S6)                     NaN            10   
38111                    Automatic (S8)                     NaN            54   

      Engine Descriptor  ...  Hours to Charge (AC 240V)  Composite City MPG  \
18087              NONE  ...                        0.0                   0   
18162         FLEX-FUEL  ...                        0.0                   0   
18189         FLEX-FUEL  ...                        0.0                   0   
18291         FLEX-FUEL  ...                        0.0                   0   
18292         FLEX-FUEL  ...                        0.0                   0   
...                 ...  ...                        ...                 ...   
38033              PHEV  ...                        0.0                  83   
38042               FFV  ...                        0.0                   0   
38054               FFV  ...                        0.0                   0   
38057               FFV  ...                        0.0                   0   
38111        SIDI; PHEV  ...                        0.0                  29   

      Composite Highway MPG Composite Combined MPG Range (FT1)  \
18087                     0                      0           0   
18162                     0                      0           0   
18189                     0                      0           0   
18291                     0    

In [182]:
df[df['Range (FT2)'].apply(lambda x: len(str(x)) > 3)]

Vehicle ID  Year    Make           Model                       Class  \
18291       16411  2000    Ford  Ranger 2WD FFV  Standard Pickup Trucks 2WD   
18292       16412  2000    Ford  Ranger 2WD FFV  Standard Pickup Trucks 2WD   
18293       16418  2000    Ford  Ranger 4WD FFV  Standard Pickup Trucks 4WD   
18294       16419  2000    Ford  Ranger 4WD FFV  Standard Pickup Trucks 4WD   
18477       16413  2000   Mazda   B3000 FFV 2WD  Standard Pickup Trucks 2WD   
...           ...   ...     ...             ...                         ...   
37378       37660  2017     GMC  Sierra K15 4WD  Standard Pickup Trucks 4WD   
37953       37992  2017     Ram        1500 2WD  Standard Pickup Trucks 2WD   
37956       37995  2017     Ram        1500 4WD  Standard Pickup Trucks 4WD   
38054       38122  2017  Toyota  Tundra 2WD FFV  Standard Pickup Trucks 2WD   
38057       38127  2017  Toyota  Tundra 4WD FFV  Standard Pickup Trucks 4WD   

                            Drive       Transmission Transmission Descriptor  \
18291            Rear-Wheel Drive  Automatic 4-Speed                   CLKUP   
18292            Rear-Wheel Drive     Manual 5-Speed                     NaN   
18293  4-Wheel or All-Wheel Drive  Automatic 4-Speed                   CLKUP   
18294  4-Wheel or All-Wheel Drive     Manual 5-Speed                     NaN   
18477            Rear-Wheel Drive  Automatic 4-Speed                   CLKUP   
...                           ...                ...                     ...   
37378               4-Wheel Drive  Automatic 6-Speed                     NaN   
37953            Rear-Wheel Drive  Automatic 8-Speed                     NaN   
37956               4-Wheel Drive  Automatic 8-Speed                     NaN   
38054            Rear-Wheel Drive     Automatic (S6)                     NaN   
38057     Part-time 4-Wheel Drive     Automatic (S6)                     NaN   

       Engine Index Engine Descriptor  ...  Hours to Charge (AC 240V)  \
18291             0         FLEX-FUEL  ...                        0.0   
18292             0         FLEX-FUEL  ...                        0.0   
18293             0         FLEX-FUEL  ...                        0.0   
18294             0         FLEX-FUEL  ...                        0.0   
18477             0         FLEX-FUEL  ...                        0.0   
...             ...               ...  ...                        ...   
37378           639         SIDI; FFV  ...                        0.0   
37953           681               FFV  ...                        0.0   
37956           690               FFV  ...                        0.0   
38054             9               FFV  ...                        0.0   
38057            10               FFV  ...                        0.0   

       Composite City MPG Composite Highway MPG Composite Combined MPG  \
18291                   0                     0                      0   
18292                   0                     0                      0   
18293                   0                     0                      0   
18294                   0                     0                      0   
18477                   0                     0                      0   
...                   ...                   ...                    ...   
37378                   0                     0                      0   
37953                   0                     0                      0   
37956                   0                     0                      0   
38054                   0                     0                      0   
38057                   0                     0                      0   

      Range (FT1) City Range (FT1) Highway Range (FT1)  Range (FT2)  \
18291           0              0.0                 0.0  230/270/270   
18292           0              0.0                 0.0  240/290/290   
18293           0              0.0                 0.0  220/270/260   
18294           0              0.0             

Additional research online revealed that the "slash" notation (e.g. 230/270/270) in the Range (FT2) column denotes differences between pickup trucks with "regular cab short wheelbase, regular cab long wheelbase, and super cab models, respectively". I found this information in a report on [FuelEconomy.gov](https://www.fueleconomy.gov/FEG/pdfs/guides/FEG2000_Part1.pdf).

To simplify this column, I think it's reasonable to take the average of the values and use that single value instead. First, though, I will replace null values with 0.

In [185]:
df['Range (FT2)'].fillna(0, inplace=True)

In [187]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Fuel Type 2                        36675    1438
Gas Guzzler Tax                    35817    2296
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [188]:
df['Fuel Type 2'].unique()

array([nan, 'Natural Gas', 'E85', 'Propane', 'Electricity'], dtype=object)

Fuel Type 2 null values can be replaced with "None". Not all vehicles will have an alternate fuel type that they can use:

In [189]:
df['Fuel Type 2'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Gas Guzzler Tax                    35817    2296
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
Manufacturer Code                  30794    7319
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [190]:
df['Gas Guzzler Tax'].unique()

array([nan, 'T', 'G', 'S'], dtype=object)

The US Department of Energy states that for the **Gas Guzzler Tax** column, vehicles with "G" or "T" values are subject to the gas guzzler tax. However, the website does not mention what "S" denotes in this case. I will follow my usual convention of converting null values to "None".

In [191]:
df['Gas Guzzler Tax'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
Manufacturer Code                  30794    7319
Transmission Descriptor            23068   15045
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]